### ISSUE TITLE
intermittently unable to open/unlock database with correct master passphrase

### ISSUE THREAD

**Issue by jflorian** [keepassxreboot/keepassxc#7249]
## Overview
With increasing frequency, I'm having problems opening and/or unlocking my database where I will see the following error despite entering the correct passphrase:
> Error while reading the database: Invalid credentials were provided, please try again.
If this reoccurs, then your database file may be corrupt. (HMAC mismatch)

## Steps to Reproduce
I'm not entirely sure, or I would likely have a fix by now.  Once I believed the database was getting corrupted because I could open the "old" variant by the built-in backup-before-save feature.  Eventually, I discovered that it wasn't (always, at least) becoming corrupted because another attempt might succeed.  Yes, this sounds like I'm simply mis-entering my master passphrase.  Believe me, I've thought the same many times, but I've carefully verified that's not the case once I started smelling a bug.  For certain I've had the following scenario many times:
1. successfully open and unlock DB and use it normally for a workday
2. lock workstation and with it, the DB
3. say next day, unlock workstation but unlocking DB gives the above error, repeatedly despite the correct passphrase
4. restart keepassxc app or reboot or ??? -- previously, this step was to restore from backup
5. possibly open and unlock DB successfully and use it normally once again

## Context
Database is kept on shared NFS storage and is accessible by several workstations and one more via sshfs.  Problem doesn't seem host-specific and I've ruled out concurrency issues having observed the problem several times where I can be assured only one workstation was accessing it.  Things that may have helped, but aren't guaranteed to do so are:

- reboot the workstation
- temporarily open a different database first, e.g., the "old" one
- open from a different workstation

I've also tried unmounting the filesystem that retains the database and remounting, but that doesn't seem to have helped.  When one host can't open/unlock the DB and another will, I've verified some OS sanity by comparing md5sum's of the same file as mounted from the two different hosts and have always found them to match.  (I've actually been keeping a checksum file for the kdbx files so I can be certain "this is the same DB that worked just yesterday".  This also helps me avoid the tedium of restoring from backup which I've done so many times before, though I'm still not entirely certain I won't have to again because I'm not yet fully convinced it isn't sometimes getting corrupted.

Most appreciated right now would be any guidance on how to best debug this further.  Should I consider this DB "tainted" somehow and proceed with a dump and import into a new one?

[NOTE]: # ( Paste debug info from Help → About here )
KeePassXC - Version 2.6.6
Revision: 9c108b9

Qt 5.15.2
Debugging mode is disabled.

Operating system: Fedora Linux 35 (KDE Plasma)
CPU architecture: x86_64
Kernel: linux 5.15.11-200.fc35.x86_64

Enabled extensions:
- Auto-Type
- Browser Integration
- SSH Agent
- KeeShare (signed and unsigned sharing)
- YubiKey
- Secret Service Integration

Cryptographic libraries:
- libgcrypt 1.9.4-unknown

[NOTE]: # ( Pick choices based on your environment )
Operating System: Linux
Desktop Env: KDE
Windowing System: X11

**Comment by droidmonkey** [comment]
All signs point to the NFS as causing your problems. Perhaps a read error in the middle of loading the database file. Can you reproduce if you copy the database to your local computer and open it from there?

The other problem we have seen, although host specific, is faulty RAM causing errors during key derivation. This is especially possible using Argon2 with large parameters that cause long derivation times.

**Comment by jflorian** [comment]
@droidmonkey Thanks for the tips.  I've come back to questioning NFS numerous times.  I did make a local copy while writing this up but don't have any history yet as to how/if that helps.  It didn't at the time, but I also got the copy via NFS so I'm going to set up a cron job to keep a local copy via scp.

The one other time I had a bizzaro case (with gzip) that made no sense whatsoever I was suggested to do a RAM check and sure enough I had faulty RAM.  I quite likely checked when I built these systems, but haven't since.  One is a Dell Precision beast with ECC, however.  I am using the default kdbx4 format with Argon2d.  If I guess at the theory of operation correctly, key derivation (which here is highly intolerant of weak RAM) happens immediately after providing a master passphrase and then that key is retained for deciphering the DB content so long as the DB remains unlocked.  Is that correct?  If so, it would seem that the well-built host with ECC wouldn't be failing the key derivation but maybe the DB was actually corrupt in that case.  I can imagine corruption happening if the derived key mutated somehow before DB changes were written out.

**Comment by droidmonkey** [comment]
We create a new random seed on every save so it is possible the new transformed key would be corrupted in some way. That would lead to permanent corruption though, not intermittent unlock failures.

**Comment by gorodoe** [comment]
Having a similar issues, turns out i have a faulty RAM (memtest showed error detection). The issues so far are completely gone once the ram was swapped out (verified by memtest that the replacement have no error). 

To replicate the issue (when still using the faulty RAM). I just create a new database, put some dummy password (also copy it into a notepad so no one accuses me of typing an incorrect password), set all the setting to default (AES-Argon2-1sec). Create a new dummy entry, then save the database. Lock/Close the application. Then try to unlock the database; anything after a save while the RAM is faulty, 100% of the time when I tried to unlock it, keepassxc will shows HMAC mismatch when tried to open them.

On the case of existing database that weren't or haven't been saved while the RAM is being faulty (somehow my RAM are not consistently faulty), opening the database is a 50-50 chance of unlocking the database, while the other 50% you get HMAC mismatch, but with multiple tries it will unlock after a few tries. If I, at a time had saved a change in the database while the RAM is being faulty, and tried to unlock the database (on a faulty RAM or another non faulty computer), the database will never unlock.

Trying to repair the database also yields no success.

While on a faulty RAM, this issue doesn't replicate when I used AES-KDF, which is given since it's not a memory hardened KDF.
IMO make a unencrypted csv backup to an encrypted container / a kdbx with AES-KDF with a different password than your master password, before trying to diagnose the issues further.

**Comment by jflorian** [comment]
I finally got around the Fedora uefi boot + memtest86+ challenge and was finally able to run a test.  I had hundreds of errors almost immediately.  Went into the BIOS config to review the "eXtreme Memory System" or whatever Gigabyte calls it on this Z97X-UD3H-BK-CF motherboard.  Anyway, this RAM (Crucial Ballistix) is supposedly spec'd as DDR3-1600 and that's what I was running things as, but it required the XMS profile to be enabled.  I turned that off and it fell back to 1333 MHz.  With that changed, I repeated the test and ran 1+ memtest86+ passes with zero errors!  Funnily enough, after having thought about this machine more, it has had quite a bit of flakiness going on, but I had attributed that to other major (many) projects going on.  It was only KeePassXC that I noticed so much because recovery was more painful, especially when the passwords I needed were those for infra projects I have going on which directly affect my ability to even reach the database filesystem.

Of course, I had to suffer at least one more DB corruption, apparently from writing the file just before the reboot to change the RAM speed.  With that cleared up however, I unlocked, added an entry, locked, unlocked again without any issue.  That's far from a thorough test, but it looks promising ... and there's the smoking gun already found.  This annoyance may have just culminated in a major relief for far much more than I ever imagined.  Yay!!!

Thanks to @droidmonkey  and @gorodoe for your valuable time and wisdom.
